be/src/storage/vectorized/delta_writer.cpp

```diff
@@ -14,43 +13,46 @@
 
 namespace starrocks::vectorized {
 
-Status DeltaWriter::open(const DeltaWriterOptions& opt, MemTracker* mem_tracker, DeltaWriter** writer) {
-    *writer = new DeltaWriter(opt, mem_tracker, StorageEngine::instance());
-    return Status::OK();
+StatusOr<std::unique_ptr<DeltaWriter>> DeltaWriter::open(const DeltaWriterOptions& opt, MemTracker* mem_tracker) {
+    std::unique_ptr<DeltaWriter> writer(new DeltaWriter(opt, mem_tracker, StorageEngine::instance()));
+    SCOPED_THREAD_LOCAL_MEM_TRACKER_SETTER(mem_tracker);
+    RETURN_IF_ERROR(writer->_init());
+    return std::move(writer);
 }
 
-DeltaWriter::DeltaWriter(const DeltaWriterOptions& opt, MemTracker* parent, StorageEngine* storage_engine)
-        : _opt(opt),
+DeltaWriter::DeltaWriter(const DeltaWriterOptions& opt, MemTracker* mem_tracker, StorageEngine* storage_engine)
+        : _state(kUninitialized),
+          _opt(opt),
+          _mem_tracker(mem_tracker),
+          _storage_engine(storage_engine),
           _tablet(nullptr),
           _cur_rowset(nullptr),
           _rowset_writer(nullptr),
+          _mem_table(nullptr),
           _tablet_schema(nullptr),
-          _delta_written_success(false),
-          _storage_engine(storage_engine) {
-    _mem_tracker = std::make_unique<MemTracker>(-1, "delta writer", parent, true);
-}
+          _flush_token(nullptr) {}
 
 DeltaWriter::~DeltaWriter() {
-    if (_is_init && !_delta_written_success) {
+    SCOPED_THREAD_LOCAL_MEM_TRACKER_SETTER(_mem_tracker);
+    switch (_get_state()) {
+    case kUninitialized:
+    case kCommitted:
+        break;
+    case kWriting:
+    case kClosed:
+    case kAborted:
         _garbage_collection();
+        break;
     }
-
     _mem_table.reset();
-
-    if (!_is_init) {
-        return;
-    }
-
-    if (_flush_token != nullptr) {
-        // cancel and wait all memtables in flush queue to be finished
-        _flush_token->cancel();
```

Comment:
Why remove _flush_token cancel when writer destructs and aborts?

Reply:
It's not required to cancel flush token before destruction